kolibri/core/discovery/utils/network/search.py

```diff
@@ -130,31 +145,75 @@ def remove_service(self, zeroconf, type, name):
             del self.instances[id]
 
 
-def get_available_instances(timeout=2, include_local=True):
+cache = Cache(os.path.join(KOLIBRI_HOME, "zeroconf_cache"))
+
+
+def gather_instance_data(instance):
+    endpoint_path = "/api/content/channel/?available=true"
+    network_client = NetworkClient(instance.get("base_url"))
+    instance["data"]["channels"] = network_client.get(endpoint_path).json()
+
+
+def find_peer_instances():
     """Retrieve a list of dicts with information about the discovered Kolibri instances on the local network,
     filtering out those that can't be accessed at the specified port (via attempting to open a socket)."""
     if not ZEROCONF_STATE["listener"]:
         initialize_zeroconf_listener()
         time.sleep(3)
-    instances = []
-    for instance in ZEROCONF_STATE["listener"].instances.values():
-        if instance["local"] and not include_local:
-            continue
-        if not _is_port_open(instance["ip"], instance["port"], timeout=timeout):
-            continue
-        instance["self"] = (
-            ZEROCONF_STATE["service"] and ZEROCONF_STATE["service"].id == instance["id"]
-        )
-        instances.append(instance)
-    return instances
+
+    return list(ZEROCONF_STATE["listener"].instances.values())
+
+
+def get_available_instances(timeout=2, include_local=True):
+    """Find peer Kolibri instances, check their availability then add them to the database.
+    Returns either a `DynamicNetworkLocation` queryset or a list of DynamicNetworkLocation objects."""
+
+    discoveries_are_fresh = cache.get(ZEROCONF_DISCOVERIES_ARE_FRESH)
+
+    if discoveries_are_fresh:
+        # a queryset pointing at already discovered addresses
+        return DynamicNetworkLocation.objects
+    else:
+
+        for instance in find_peer_instances():
+            if instance["local"] and not include_local:
+                continue
+
+            if instance["self"]:
+                continue
+
+            try:
+                (
+                    network_location,
+                    created,
+                ) = DynamicNetworkLocation.objects.update_or_create(
```

Comment:
here's where we cache the discovered peers... they will magically be in the same format as the manually saved ones, so that the content import wizard doesn't have to know that these dynamically discovered peers are any different from manually saved ones.